# **- Machine Learning -**
# _* **Boosting** with Python *_
## **+ '``Gradient Boosting``' , '``Adaptive Boosting``' and ``'Extreme Gradient Booster'`` +**
### **+ _Diabetes Data - Hitters Data_ +**

In [1]:
%matplotlib inline

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import accuracy_score
from sklearn.metrics import r2_score
from sklearn.model_selection import GridSearchCV
from sklearn import preprocessing
from sklearn.pipeline import Pipeline

**Learning Rate (LAMBDA)** ile **kaç tane Ağaç (B)** kuracağımızı belirlememiz lazım... Bunlar bizim için birer Parametre... Bu Parametreleri belirlemek için de **Cross Validation** yapmamız lazım... Yani en ideal **Boosting**'i bulmamız lazım... Bunun için de **GridSearchCV** kullanmamız lazım... 

# ``Classification`` with **Diabetes Data :**

In [2]:
df = pd.read_table('diabetesdata.txt')
X = df.drop('Diabetes',axis=1)
y = df['Diabetes']
X_col = X.columns
df

Preg  Glucose  BP  SkinThick  Insul   BMI    DPF  Age  Diabetes
0       6      148  72         35      0  33.6  0.627   50         1
1       1       85  66         29      0  26.6  0.351   31         0
2       8      183  64          0      0  23.3  0.672   32         1
3       1       89  66         23     94  28.1  0.167   21         0
4       0      137  40         35    168  43.1  2.288   33         1
..    ...      ...  ..        ...    ...   ...    ...  ...       ...
763    10      101  76         48    180  32.9  0.171   63         0
764     2      122  70         27      0  36.8  0.340   27         0
765     5      121  72         23    112  26.2  0.245   30         0
766     1      126  60          0      0  30.1  0.349   47         1
767     1       93  70         31      0  30.4  0.315   23         0

[768 rows x 9 columns]

In [3]:
X_col

Index(['Preg', 'Glucose', 'BP', 'SkinThick', 'Insul', 'BMI', 'DPF', 'Age'], dtype='object')

In [4]:
# Split data into training and test sets ::

X_train , X_test , y_train , y_test = train_test_split(X, y, test_size=0.2, random_state=1,stratify=y)

## **Gradient Boosting**
### **``Classicifation``** with ``Gradient Boosting``
#### ``GradientBoostingClassifier()`` modülümüzün içerisinde neler varmış bakalım **:**

In [5]:
from sklearn.ensemble import GradientBoostingClassifier

GradientBoostingClassifier()

GradientBoostingClassifier()

+ + DEFAULT olarak **Ağaç** oluşturuyor... Yani bir **Ağaç** FIT ettiriyor...
+ + DEFAULT olarak **(max_depth=3)** Parametresi **:** Normalde **Decision Tree**'de Ağaç kurarken DEFAULT olarak **maksimum derinlik sınırı** yoktu... Ama **Gradient Boosting Classifier**'de bu var... Yani daha derin **Ağaçlar** yaratmıyor... Yani çok da iyi öğrenebilen **Ağaçlar** kurmuyor... Bunun **Ağaçları** biraz salak... Tabii bu değeri değiştirebiliriz...
+ + DEFAULT olarak **(n_estimators=100)** Parametresi **:** Default olarak **'100'** tane **Ağaç ,** yani **STUMP** FIT ediyor...
+ + DEFAULT olarak **(learning_rate=0.1)** Parametresi **:** Bu değer LAMBDA'dır... Ne kadar küçük olursa **,** modelin **her bir Ağacının** Tahmini o kadar küçülür **,** değersizleşir...

+ Şimdi de bir **GradientBoostingClassifier()** modeli kuralım **:**

In [6]:
model = GradientBoostingClassifier(n_estimators=1000 , max_depth=1 , learning_rate=1)

In [7]:
model.fit(X_train , y_train)

y_pred = model.predict(X_test)

print('Accuracy:' , accuracy_score(y_test, y_pred))

Accuracy: 0.6818181818181818


Yukarıdaki **Accuracy** değerine baktığımızda **,** çok da başarılı bir sonuç aldığımız söylenemez...

#### There is a trade-off between number of trees and learning rate.
#### **Ağaç sayısı** ile **öğrenme hızı (oranı)** arasında bir değiş tokuş vardır.
#### You should never check the test performance when tuning parameters!!! This is only for illustrative purposes
#### Parametreleri ayarlarken , asla test performansını kontrol etmemelisiniz!!!

+ **Learning Rate**'yi biraz daha azaltalım. (DEFAULT hale getirelim) **:**
<br> Yani biraz daha YAVAŞ öğrenelim...

In [8]:
model = GradientBoostingClassifier(n_estimators=1000 , max_depth=1 , learning_rate=0.1)

model.fit(X_train , y_train)

y_pred = model.predict(X_test)

print('Accuracy:' , accuracy_score(y_test, y_pred))

Accuracy: 0.7467532467532467


Yukarıda gördüğümüz gibi **, Learning Rate**'yi azaltınca **Accuracy** değeri de anlamlı bir şekilde arttı...

+ Yukarıdaki iki örnekte Ağaç Sayısı yani iterasyon sayısı birbirleri ile AYNI olmasına rağmen **,** Öğrenme Hızı'nı **azaltınca** neden Accuracy değeri arttı **?** Veya Öğrenme Hızı'nı **arttırınıca** neden Accuracy değeri düştü **?**
+ + AYNI sayıda Ağaç varken **,** Öğrenme Hızı'nı arttırınca OVERFIT etme ihtimalimiz de artar. Hızlıca her şeyi öğrenmeye çalıştığı için muhtemelen OVERFIT oldu ve onun için **Accuracy** değeri daha düşük çıktı...

+ Şimdi de **Learning Rate**'yi biraz daha düşürelim **:**

In [6]:
model = GradientBoostingClassifier(n_estimators=1000 , max_depth=1 , learning_rate=0.01)

model.fit(X_train , y_train)

y_pred = model.predict(X_test)

print('Accuracy:' , accuracy_score(y_test, y_pred))

Accuracy: 0.7337662337662337


Yukarıda gördüğümüz gibi **, Öğrenme Hızı**'nı daha da azaltmamıza RAĞMEN BU DEFA **Accuracy** değeri biraz düştü... Şimdi de çok büyük ihtimalle UNDERFIT oldu. Yani o kadar yavaş öğrenmeye çalıştı ki **,** öğrenebileceği bazı şeyleri öğrenemedi...

In [7]:
model = GradientBoostingClassifier(n_estimators=10000 , max_depth=1 , learning_rate=0.01)
model.fit(X_train , y_train)
y_pred = model.predict(X_test)
print('Accuracy:' , accuracy_score(y_test, y_pred))

Accuracy: 0.7467532467532467


**++++ CRITICAL NOTE ++++**
<br> İdeali **;** olabildiğince yavaş öğrenip **,** bir o kadar da çok modele sormaktır... Yani **'learning_rate'**'yi olabildiğince azaltıp **, 'n_estimators'**'ü olabildiğince arttırmaktır...
<br> Ama her zaman da böyle olacak diye bir şey yok... Onun için en ideal Parametreleri bir şekilde bulmamız gerelir...

### Parameter Tuning

Şimdi de **,** ideal Parametreleri bulmaya çalışalım **:**

In [8]:
model = GradientBoostingClassifier(max_depth=1)

# Number of trees fit ::
n_estimators = [100 , 1000 , 5000 , 10000]
# Learning rate ::
learning_rate = [1 , 0.1 , 0.01 , 0.001]
# Create grid ::
params = {
 'n_estimators' : n_estimators ,
 'learning_rate' : learning_rate ,
 }
# Random search of parameters ::
boost_grid = GridSearchCV(estimator=model , param_grid=params , 
                          cv=5 , verbose=2 , scoring='accuracy' , n_jobs= -1)
# Fit the model ::
boost_grid.fit(X_train , y_train)
# Print results ::
print(boost_grid.best_params_)

Fitting 5 folds for each of 16 candidates, totalling 80 fits
{'learning_rate': 0.01, 'n_estimators': 5000}


Yukarıda modeli kurarken **GradientBoostingClassifier(max_depth=1)** şeklinde kurduk... Yani **maksimum derinlik** Parametresini direkt yazdık... Ama istersek onu da **,** en ideali aranan Parametreler arasına ekleyip o şekilde çalıştırabilirdik...

En ideal Parametreleri **; " { 'learning_rate': 0.01 , 'n_estimators': 5000 } "** şeklinde buldu...
+ Şimdi de bu Parametreleri yerine koyup modeli EN BAŞTAN kuralım ve Tahminleme yapıp Score'nine bakalım **:**

In [9]:
model = GradientBoostingClassifier(n_estimators=5000 , max_depth=1 , learning_rate=0.01)

model.fit(X_train , y_train)

y_pred = model.predict(X_test)

print('Accuracy:' , accuracy_score(y_test, y_pred))

Accuracy: 0.7402597402597403


## **Adaptive Boosting**
Veri Noktalarının Ağırlıklarını değiştirerek çalışıyor...
Her STUMP'ta Yanlış Sınıfta kalan Veri Noktalarının Ağırlıklarını ARTTIRIYOR **,** Doğru Sınıfta kalan Veri Noktalarının da Ağırlıklarını AZALTIYOR... Bunu **,** her yeni STUMP'ta güncelleyerek devam ediyor...
### **``AdaboostClassifier``**

In [11]:
from sklearn.ensemble import AdaBoostClassifier

model = AdaBoostClassifier()

n_estimators = [50 , 100 , 1000 , 5000]
# Number of features at every split ::
learning_rate = [1 , 0.1 , 0.01 , 0.001]

# Create grid ::
params = {
 'n_estimators' : n_estimators ,
 'learning_rate' : learning_rate ,
 }
# Random search of parameters ::
boost_grid = GridSearchCV(estimator=model , param_grid=params , 
                          cv=5 , verbose=2 , scoring='accuracy' , n_jobs = -1)
# Fit the model ::
boost_grid.fit(X_train , y_train)
# Print results ::
print(boost_grid.best_params_)

Fitting 5 folds for each of 16 candidates, totalling 80 fits
{'learning_rate': 0.1, 'n_estimators': 100}


In [12]:
model = AdaBoostClassifier(n_estimators=100 , learning_rate=0.1)
model.fit(X_train , y_train)
y_pred = model.predict(X_test)
print('Accuracy:' , accuracy_score(y_test , y_pred))

Accuracy: 0.7142857142857143


## Extreme Gradient Booster (``XGBClassifier``)
#### ``Extreme Gradient Booster:`` A popular library that implements boosting faster and generally provides better performance.
#### ``Extreme Gradient Booster:`` Boosting'i daha hızlı uygulayan ve genellikle daha iyi performans sağlayan popüler bir Kütüphanedir...
<br> **Scikit-Learn**'deki **Gredient Boosting Classifier**'in bir alternatifidir...
<br> Genelde **Scikit-Learn**'den daha iyi bir performans gösterir...
<br> Ayrıntılarını , arka plandaki çalışma şeklini ve optimizasyonlarını ayrıntılı bir şekilde okuyarak öğrenebiliriz...
<br> Yine iyi çalışan başka Kütüphaneler de var... **``'LightGBM'``** Kütüphanesi de bunlardan biridir ve o da iyi çalışır...

In [13]:
#   pip install XGBoost

from xgboost import XGBClassifier

XGBClassifier()

XGBClassifier(base_score=None, booster=None, colsample_bylevel=None,
              colsample_bynode=None, colsample_bytree=None,
              enable_categorical=False, gamma=None, gpu_id=None,
              importance_type=None, interaction_constraints=None,
              learning_rate=None, max_delta_step=None, max_depth=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              n_estimators=100, n_jobs=None, num_parallel_tree=None,
              predictor=None, random_state=None, reg_alpha=None,
              reg_lambda=None, scale_pos_weight=None, subsample=None,
              tree_method=None, validate_parameters=None, verbosity=None)

In [14]:
model = XGBClassifier()

n_estimators = [50 , 100 , 1000 , 5000]
# Number of features at every split ::
learning_rate = [1 , 0.1 , 0.01 , 0.001]
# Create grid ::
params = {
 'n_estimators' : n_estimators ,
 'learning_rate' : learning_rate ,
 }
# Random search of parameters ::
boost_grid = GridSearchCV(estimator=model , param_grid=params , 
                          cv=5 , verbose=2 , scoring='accuracy' , n_jobs = -1)
# Fit the model ::
boost_grid.fit(X_train , y_train)
# Print results ::
print(boost_grid.best_params_)

Fitting 5 folds for each of 16 candidates, totalling 80 fits


C:\anaconda3\envs\ITUmindset\lib\site-packages\xgboost\sklearn.py:1224: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[05:23:30] WARNING: D:\bld\xgboost-split_1645118015404\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
{'learning_rate': 0.01, 'n_estimators': 100}


In [15]:
model = XGBClassifier(n_estimators=100 , learning_rate=0.01)

model.fit(X_train , y_train)

y_pred = model.predict(X_test)

print('Accuracy:' , accuracy_score(y_test, y_pred))

[05:27:34] WARNING: D:\bld\xgboost-split_1645118015404\work\src\learner.cc:1115: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'binary:logistic' was changed from 'error' to 'logloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
Accuracy: 0.7532467532467533


# ``Regression`` with **Hitters Data :**
## **``Regression``** with ``Gradient Boosting``
### **'GradientBoostingRegressor'**

+ Öncelikle **, Hitters** Veri Setini çağıralım ve düzenleyelim **:**
<br> **+--- NOTE ---> 'NaN'** değerleri Veri Setimizden direkt atmak yerine düzenlememiz daha doğru olacaktır... Ama burada uğraşmayacağız...

In [16]:
df = pd.read_csv('Hitters_Data.csv')

df = df.dropna()

dummies = pd.get_dummies(df[['League' , 'Division' , 'NewLeague']])

y = np.log(df.Salary)

# Drop the column with the independent variable (Salary), and columns for which we created dummy variables ::
X__ = df.drop(['Salary' , 'League' , 'Division' , 'NewLeague'] , axis = 1).astype('float64')

# Define the feature set X. ::
X = pd.concat([X__ , dummies[['League_N', 'Division_W', 'NewLeague_N']]] , axis = 1)


In [17]:
X

AtBat   Hits  HmRun  Runs   RBI  Walks  Years  CAtBat   CHits  CHmRun  \
1    315.0   81.0    7.0  24.0  38.0   39.0   14.0  3449.0   835.0    69.0   
2    479.0  130.0   18.0  66.0  72.0   76.0    3.0  1624.0   457.0    63.0   
3    496.0  141.0   20.0  65.0  78.0   37.0   11.0  5628.0  1575.0   225.0   
4    321.0   87.0   10.0  39.0  42.0   30.0    2.0   396.0   101.0    12.0   
5    594.0  169.0    4.0  74.0  51.0   35.0   11.0  4408.0  1133.0    19.0   
..     ...    ...    ...   ...   ...    ...    ...     ...     ...     ...   
317  497.0  127.0    7.0  65.0  48.0   37.0    5.0  2703.0   806.0    32.0   
318  492.0  136.0    5.0  76.0  50.0   94.0   12.0  5511.0  1511.0    39.0   
319  475.0  126.0    3.0  61.0  43.0   52.0    6.0  1700.0   433.0     7.0   
320  573.0  144.0    9.0  85.0  60.0   78.0    8.0  3198.0   857.0    97.0   
321  631.0  170.0    9.0  77.0  44.0   31.0   11.0  4908.0  1457.0    30.0   

     CRuns   CRBI  CWalks  PutOuts  Assists  Errors  League_N  Division_W  \
1    321.0  414.0   375.0    632.0     43.0    10.0         1           1   
2    224.0  266.0   263.0    880.0     82.0    14.0         0           1   
3    828.0  838.0   354.0    200.0     11.0     3.0         1           0   
4     48.0   46.0    33.0    805.0     40.0     4.0         1           0   
5    501.0  336.0   194.0    282.0    421.0    25.0         0           1   
..     ...    ...     ...      ...      ...     ...       ...         ...   
317  379.0  311.0   138.0    325.0      9.0     3.0         1           0   
318  897.0  451.0   875.0    313.0    381.0    20.0         0           0   
319  217.0   93.0   146.0     37.0    113.0     7.0         0           1   
320  470.0  420.0   332.0   1314.0    131.0    12.0         0           0   
321  775.0  357.0   249.0    408.0      4.0     3.0         0           1   

     NewLeague_N  
1              1  
2              0  
3              1  
4              1  
5              0  
..           ...  
317            1  
318            0  
319            0  
320            0  
321            0  

[263 rows x 19 columns]

#### **++++ CRUCIAL NOTE ++++**
**Regression** problemlerinde **,** öncelikle MUHAKKAK STANDARDİZASYON yapmamız gerekiyor...
<br> **Classification** problemlerinde de STANDARDİZASYON yapılabilir... Ancak **Regression** problemlerinde KESİN yapılmalıdır...

+ Veri Setimizi TRAIN ve TEST olmak üzere ayıralım ve SONRA da STANDARDİZASYON işlemlerini gerçekleştirelim **:**

In [18]:
#   Split data into training and test sets :

X_train , X_test , y_train , y_test = train_test_split(X , y , test_size=0.2 , random_state=1)

scaler = preprocessing.StandardScaler()

scaler.fit(X_train)

X_trainStandard = scaler.transform(X_train)
X_testTransformed = scaler.transform(X_test)

#### **++++ CRUCIAL NOTE ++++**
**Random Forest** konusunda da söylediğimiz gibi **;** En iyi Parametreleri bulmak için kullandığımız mesela aşağıdaki **'GradientBoostingRegressor'** tekniğine **,** yukarıdaki **Standardize Edilmiş** Veri Seti üzerinden **'Grid Search'** YAPMAMAMIZ LAZIM **!** Bu **Standardize Edilmiş** Veri Seti üzerinden **'Grid Search'** yaparsak **, VALIDATION Setini de görerek bir eğitim işlemi yapmış oluruz... Ve bu büyük bir hatadır...**
<br> Bunun için de **; 'Grid Search'**'e başlarken **, STANDARDİZE EDİLMEMİŞ Veri Seti üzerinden işlemimizi gerçekleştiriyoruz...**

AMA **PIPELINE** işleminde **;** ÖNCE STANDARDIZE Edip DAHA SONRA Tekniğimizi uyguluyoruz...

**+--- NOTE --->** Aynı şekilde **,** bu işlemlerin hepsini **CLASSIFICATION** problemlerinde de uygulayabilirdik...

In [19]:
from sklearn.ensemble import GradientBoostingRegressor

model = GradientBoostingRegressor(max_depth=1)

n_estimators = [100 , 1000 , 5000 , 10000]
# Number of features at every split ::
learning_rate = [1 , 0.1 , 0.01 , 0.001]
# Create grid ::
params = {
 'Regressor__n_estimators' : n_estimators ,
 'Regressor__learning_rate' : learning_rate ,
 }

pipe = Pipeline([('scaler', preprocessing.StandardScaler()) , ('Regressor', model)])
# Random search of parameters ::
boost_grid = GridSearchCV(estimator=pipe , param_grid=params , 
                          cv=5 , verbose=2 , scoring='neg_mean_squared_error' , n_jobs= -1)
# Fit the model ::
boost_grid.fit(X_train , y_train)
# Print results ::
print(boost_grid.best_params_)

Fitting 5 folds for each of 16 candidates, totalling 80 fits
{'Regressor__learning_rate': 0.01, 'Regressor__n_estimators': 1000}


#### **++++ CRUCIAL NOTE ++++**
Yukarıda **, PIPELINE** içerisinde hem **'scaler'** hem de **'Regressor'** Teknikleri olduğu için **;** Sözlük türünde oluşturduğumuz Parametrelerin isimlerine direkt **"'n_estimators'"** ve **"'learning_rate'"** şeklinde yazamıyoruz... **PIPELINE** içindeki hangi Tekniğe AİT İSE **,** onun ismini de MUHAKKAK belirtmemiz gerekiyor...
<br> **"'Regressor__n_estimators'"** ve **"'Regressor__learning_rate'"** ... gibi ...

In [17]:
boost_grid.best_score_

-0.2477818958088657

**++++ NOTE ++++**
<br> Yukarıdaki En İyi Parametreleri bulma işleminde **,** en iyi Skoru **,** EKSİLİ bir değer bulmasının sebebi **, (scoring='neg_mean_squared_error')** metriğini kullanmamızdır... Normalde **,** direkt **'mean_squared_error'** Skoruna baktığımızda **,** metrik Skoru EN KÜÇÜK olan Parametreyi seçiyoruz... Ama burada **'neg_mean_squared_error'** metriğinin Skoruna baktığımız için **,** bu defa da metrik skoru EN BÜYÜK olan Parametreyi seçiyoruz... Çünkü sonuçları EKSİLİ şekilde veriyor... Ama tabii bu işlemleri bizim için **Scikit-Learn** yapıyor...
<br> İstersek bu ortaya çıkan BEST SCORE'yi tekrar EKSİ ile çarpıp **,** ARTI hale getirebiliriz ama gerek yok... Biz zaten sadece En İyi Parametreyi arıyoruz...
<br> Yani **Cross Validation** değeri en iyi olan Parametrenin Skoru **'-0.2477818958088657'** imiş...

Bütün Parametre Skore'lerini görmek istersek de **, ``'boost_grid.cv_results_'``** komutunu yazabiliriz...

+ Şimdi de **,** en iyi parametreleri kullanarak EN BAŞTAN modelimizi kuralım **:**

In [20]:
model = GradientBoostingRegressor(n_estimators=1000 , max_depth=1 , learning_rate=0.01)
model.fit(X_trainStandard , y_train)
y_pred = model.predict(X_testTransformed)
print('Mean Squared Error:' , mean_squared_error(y_test, y_pred))

Mean Squared Error: 0.17972168739765562


#### **++++ CRUCIAL NOTE ++++**
Yukarıda gördüğümüz gibi **,** En İyi Parametreleri seçtikten sonra **,** modeli EN BAŞTAN kurarken ve Tahminleme yaparken işlemlerimizi **,** en başta ki **Standardize Edilmiş** Veri Seti üzerinden yapıyoruz...

In [22]:
r2_score(y_test , y_pred)

0.8011813996070491

## **``Regression``** with ``Adaptive Boosting``
### **'AdaboostRegressor'**

In [21]:
from sklearn.ensemble import AdaBoostRegressor

model = AdaBoostRegressor()

n_estimators = [100 , 1000 , 5000 , 10000]
# Number of features at every split ::
learning_rate = [1 , 0.1 , 0.01 , 0.001]
# Create grid ::
params = {
 'Regressor__n_estimators' : n_estimators ,
 'Regressor__learning_rate' : learning_rate ,
 }

pipe = Pipeline([('scaler', preprocessing.StandardScaler()) , ('Regressor', model)])
# Random search of parameters ::
boost_grid = GridSearchCV(estimator=pipe , param_grid=params , 
                          cv=5 , verbose=2 , scoring='neg_mean_squared_error' , n_jobs = -1)
# Fit the model ::
boost_grid.fit(X_train , y_train)
# Print results ::
print(boost_grid.best_params_)

Fitting 5 folds for each of 16 candidates, totalling 80 fits
{'Regressor__learning_rate': 1, 'Regressor__n_estimators': 100}


In [25]:
model = AdaBoostRegressor(n_estimators=100 , learning_rate= 1)
model.fit(X_trainStandard , y_train)
y_pred = model.predict(X_testTransformed)
print('Mean Squared Error:' , mean_squared_error(y_test, y_pred))

Mean Squared Error: 0.1949846603112319


## **``Regression``** with ``Extreme Gradient Booster`` (**XGBoostRegressor**)
### **'XGBRegressor'**

In [27]:
from xgboost import XGBRegressor

model = XGBRegressor()

n_estimators = [100 , 1000 , 5000 , 10000]
# Number of features at every split ::
learning_rate = [1 , 0.1 , 0.01 , 0.001]
# Create grid ::
params = {
 'Regressor__n_estimators' : n_estimators ,
 'Regressor__learning_rate' : learning_rate ,
 }

pipe = Pipeline([('scaler', preprocessing.StandardScaler()) , ('Regressor', model)])
# Random search of parameters ::
boost_grid = GridSearchCV(estimator=pipe , param_grid=params , 
                          cv=5 , verbose=2 , scoring='neg_mean_squared_error' , n_jobs = -1)
# Fit the model ::
boost_grid.fit(X_train , y_train)
# Print results ::
print(boost_grid.best_params_)

Fitting 5 folds for each of 16 candidates, totalling 80 fits
{'Regressor__learning_rate': 0.01, 'Regressor__n_estimators': 5000}


In [28]:
model = XGBRegressor(n_estimators=5000 , learning_rate=0.01)
model.fit(X_trainStandard , y_train)
y_pred = model.predict(X_testTransformed)
print('Mean Squared Error:' , mean_squared_error(y_test, y_pred))

Mean Squared Error: 0.193993561571042


In [31]:
model.feature_importances_

array([0.01663682, 0.00632886, 0.00481615, 0.01029918, 0.00993644,
       0.01325134, 0.02696959, 0.23648155, 0.36681116, 0.06597257,
       0.08040313, 0.08907861, 0.03242335, 0.01212592, 0.00587025,
       0.0079202 , 0.00609087, 0.00444641, 0.00413751], dtype=float32)

**``'feature_importances_'``** komutu **;** HER BİR FEATURE'nin ne kadar **'IMPORTANT (ÖNEMLİ)'** olduğunu gösterir...
<br> En iyi model kurulduktan sonra **,** sonuç olarak çıkan bu değerler **ne kadar büyük ise metrik sonucu o kadar küçülmüş demektir...** Yani iyileşmiş demektir... Onun için de en önemli Feature **,** EN ÇOK **improvement** olan **,** yani iyileşme gösteren Feature'dir...